File: go/problem_50.go
package main

type ExprNode struct {
	val         string
	left, right *ExprNode
}

func evalExpr(root *ExprNode) float64 {
	if root.left == nil && root.right == nil {
		v := 0.0
		for _, c := range root.val {
			v = v*10 + float64(c-'0')
		}
		return v
	}
	l := evalExpr(root.left)
	r := evalExpr(root.right)
	switch root.val {
	case "+":
		return l + r
	case "-":
		return l - r
	case "*":
		return l * r
	default:
		return l / r
	}
}

func problem50() {
	println("Problem 50:")
	root := &ExprNode{val: "*",
		left:  &ExprNode{val: "+", left: &ExprNode{val: "3"}, right: &ExprNode{val: "2"}},
		right: &ExprNode{val: "+", left: &ExprNode{val: "4"}, right: &ExprNode{val: "5"}},
	}
	if evalExpr(root) != 45 {
		panic("assertion failed")
	}
	println("All Tests Passed")
}

File: python/problem_50.py
class ExprNode:
    def __init__(self, val, left=None, right=None):
        self.val = val; self.left = left; self.right = right


def eval_expr(root: ExprNode) -> float:
    if root.left is None and root.right is None:
        return float(root.val)
    l = eval_expr(root.left)
    r = eval_expr(root.right)
    if root.val == '+': return l + r
    if root.val == '-': return l - r
    if root.val == '*': return l * r
    return l / r


def main():
    print("Problem 50:")
    root = ExprNode('*',
        ExprNode('+', ExprNode('3'), ExprNode('2')),
        ExprNode('+', ExprNode('4'), ExprNode('5')),
    )
    assert eval_expr(root) == 45.0
    print("All Tests Passed")
